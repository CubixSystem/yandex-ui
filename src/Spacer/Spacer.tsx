import React, { FC, RefObject, CSSProperties, ElementType } from 'react';
import { cn } from '@bem-react/classname';

type SpacerSharedProps = {
    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: RefObject<HTMLElement>;

    /**
     * Пользовательские стили
     */
    style?: CSSProperties;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Компонент для отображения
     *
     * @default 'div'
     */
    as?: ElementType;
};

type SpacerKindProps =
    | {
          all?: string | number;
          vertical?: never;
          horizontal?: never;
          top?: never;
          bottom?: never;
          left?: never;
          right?: never;
      }
    | {
          vertical?: string | number;
          horizontal?: string | number;
          all?: never;
          top?: never;
          bottom?: never;
          left?: never;
          right?: never;
      }
    | {
          all?: never;
          vertical?: never;
          horizontal?: never;
          top?: string | number;
          bottom?: string | number;
          left?: string | number;
          right?: string | number;
      };

export type SpacerProps = SpacerSharedProps & SpacerKindProps;

export const cnSpacer = cn('Spacer');

const getSpace = ({ all, vertical, horizontal, top, left, bottom, right }: any): any => {
    if (all !== undefined) {
        return {
            padding: all,
        };
    }
    if (vertical !== undefined || horizontal !== undefined) {
        return {
            paddingTop: vertical,
            paddingBottom: vertical,
            paddingLeft: horizontal,
            paddingRight: horizontal,
        };
    }
    if (top !== undefined || left !== undefined || bottom !== undefined || right !== undefined) {
        return {
            paddingTop: top,
            paddingBottom: bottom,
            paddingLeft: left,
            paddingRight: right,
        };
    }
};

/**
 * Компонент который вставляет своих детей с заданным смещением.
 * @param {SpacerProps} props
 */
export const Spacer: FC<SpacerProps> = ({
    all,
    vertical,
    horizontal,
    top,
    bottom,
    left,
    right,
    innerRef,
    style,
    className,
    children,
    as: Component = 'div',
    ...props
}) => (
    <Component
        {...props}
        ref={innerRef}
        className={cnSpacer(null, [className])}
        style={{ ...getSpace({ all, vertical, horizontal, top, left, bottom, right }), ...style }}
    >
        {children}
    </Component>
);

Spacer.displayName = cnSpacer();
