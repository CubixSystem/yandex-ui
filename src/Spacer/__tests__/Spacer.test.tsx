import React, { createRef } from 'react';

import { createClientRender, screen } from '../../internal/testing';
import { Spacer } from '../Spacer';

describe('Spacer', () => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(<Spacer className="mix" />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен быть установлен корректный displayName', () => {
        expect(Spacer.displayName).toBe('Spacer');
    });

    test('должен устанавливать ссылку на корневой DOM элемент', async () => {
        const innerRef = createRef<HTMLDivElement>();
        render(<Spacer innerRef={innerRef} />);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен устанавливать и изменять all через props', () => {
        const { setProps } = render(<Spacer data-testid="element" all={20} />);

        expect(screen.getByTestId('element')).toHaveStyle({
            padding: '20px',
        });

        setProps({ all: '10rem' });

        expect(screen.getByTestId('element')).toHaveStyle({
            padding: '10rem',
        });
    });

    test('должен устанавливать vertical и horizontal через props', () => {
        render(<Spacer data-testid="element" vertical={10} horizontal={20} />);

        expect(screen.getByTestId('element')).toHaveStyle({
            paddingBottom: '10px',
            paddingLeft: '20px',
            paddingRight: '20px',
            paddingTop: '10px',
        });
    });

    test('должен устанавливать top bottom left и right через props', () => {
        render(<Spacer data-testid="element" top={10} bottom="10px" left="20rem" right={20} />);

        expect(screen.getByTestId('element')).toHaveStyle({
            paddingBottom: '10px',
            paddingLeft: '20rem',
            paddingRight: '20px',
            paddingTop: '10px',
        });
    });

    test('должен изменять тег компонента через props as', () => {
        render(<Spacer as="span" data-testid="element" />);

        expect(screen.getByTestId('element').tagName).toBe('SPAN');
    });
});
