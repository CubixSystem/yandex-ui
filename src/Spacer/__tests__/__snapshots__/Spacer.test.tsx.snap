// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spacer должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Spacer mix"
/>
`;
